staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1467 @@
+// connection agent, or (2) it's part of a Virtual Network where at least one
+// NetworkAttachment for which (1) is true exists. To achieve its goal, a
+// connection agent receives notifications about relevant NetworkAttachments
+// from the K8s API server through Informers, and when necessary
+// creates/updates/deletes Network Interfaces through a low-level network
+// interface fabric. When a new Virtual Network becomes relevant for the
+// connection agent because of the creation of the first attachment of that
+// Virtual Network on the same node as the connection agent, a new informer on
+// remote NetworkAttachments in that Virtual Network is created. Upon being
+// notified of the creation of a local NetworkAttachment, the connection agent
+// also updates the status of such attachment with its host IP and the name of
+// the interface which was created.
+type ConnectionAgent struct {
+	localNodeName string
+	hostIP        gonet.IP
+	kcs           *kosclientset.Clientset
+	netv1a1Ifc    netvifc1a1.NetworkV1alpha1Interface
+	eventRecorder k8seventrecord.EventRecorder
+	queue         k8sworkqueue.RateLimitingInterface
+	workers       int
+	netFabric     netfabric.Interface
+	stopCh        <-chan struct{}
+
+	// Informer and lister on NetworkAttachments on the same node as the
+	// connection agent
+	localAttsInformer k8scache.SharedIndexInformer
+	localAttsLister   koslisterv1a1.NetworkAttachmentLister
+
+	// Map from vni to vnState associated with that vni. Accessed only while
+	// holding vniToVnStateMutex
+	vniToVnStateMutex sync.RWMutex
+	vniToVnState      map[uint32]*vnState
+
+	// nsnToVNStateVNI maps local attachments namespaced names to the VNI of the
+	// vnState they're stored in. Accessed only while holding nsnToVNStateVNIMutex.
+	nsnToVNStateVNIMutex sync.RWMutex
+	nsnToVNStateVNI      map[k8stypes.NamespacedName]uint32
+
+	nsnToLocalStateMutex      sync.RWMutex
+	nsnToLocalMainState       map[k8stypes.NamespacedName]LocalAttachmentMainState
+	nsnToPostCreateExecReport map[k8stypes.NamespacedName]*netv1a1.ExecReport
+
+	nsnToRemoteIfcMutex sync.RWMutex
+	nsnToRemoteIfc      map[k8stypes.NamespacedName]netfabric.RemoteNetIfc
+
+	// nsnToVNIs maps attachments (both local and remote) namespaced names
+	// to set of vnis where the attachments have been seen. It is accessed by the
+	// notification handlers for remote attachments, which add/remove the vni
+	// with which they see the attachment upon creation/deletion of the attachment
+	// respectively. When a worker processes an attachment reference, it reads
+	// from nsnToVNIs the vnis with which the attachment has been seen. If there's
+	// more than one vni, the current state of the attachment is ambiguous and
+	// the worker stops the processing. The deletion notification handler for one
+	// of the VNIs of the attachment will cause the reference to be requeued
+	// and hopefully by the time it is dequeued again the ambiguity as for the
```

Comment:
The delete in the notification has already taken effect in the informer's cache when the notification is delivered.  Of course, other things can happen in the meantime.  What is certain is that the set of VNIs for a given NSN will reflect more recent authoritative states.

Reply:
Yes. The point is that "hopefully ... the ambiguity ... has been resolved" is a weaker statement than we can make.

Reply:
Mmm... I think you're assuming that there are only two VNIs. If there are more than two, I think "hopefully ... the ambiguity ... has been resolved" is all we can guarantee. But probably a better rephrasing will make that clearer.